include/deal.II/base/vectorization.h

```diff
@@ -825,6 +1061,176 @@ class VectorizedArray<float>
 
 
 
+/**
+ * Specialization for float and AVX-512.
+ */
+template <>
+inline
+void
+vectorized_load_and_transpose(const unsigned int      n_entries,
+                              const float            *in,
+                              const unsigned int     *offsets,
+                              VectorizedArray<float> *out)
+{
+  const unsigned int n_chunks = n_entries/4, remainder = n_entries%4;
+  for (unsigned int outer = 0; outer<16; outer += 8)
+    {
+      for (unsigned int i=0; i<n_chunks; ++i)
+        {
+          __m128 u0 = _mm_loadu_ps(in+4*i+offsets[0+outer]);
+          __m128 u1 = _mm_loadu_ps(in+4*i+offsets[1+outer]);
+          __m128 u2 = _mm_loadu_ps(in+4*i+offsets[2+outer]);
+          __m128 u3 = _mm_loadu_ps(in+4*i+offsets[3+outer]);
+          __m128 u4 = _mm_loadu_ps(in+4*i+offsets[4+outer]);
+          __m128 u5 = _mm_loadu_ps(in+4*i+offsets[5+outer]);
+          __m128 u6 = _mm_loadu_ps(in+4*i+offsets[6+outer]);
+          __m128 u7 = _mm_loadu_ps(in+4*i+offsets[7+outer]);
+          // To avoid warnings about uninitialized variables, need to initialize
+          // one variable with zero before using it.
+          __m256 t0, t1, t2, t3 = _mm256_set1_ps(0.F);
+          t0 = _mm256_insertf128_ps (t3, u0, 0);
+          t0 = _mm256_insertf128_ps (t0, u4, 1);
+          t1 = _mm256_insertf128_ps (t3, u1, 0);
+          t1 = _mm256_insertf128_ps (t1, u5, 1);
+          t2 = _mm256_insertf128_ps (t3, u2, 0);
+          t2 = _mm256_insertf128_ps (t2, u6, 1);
+          t3 = _mm256_insertf128_ps (t3, u3, 0);
+          t3 = _mm256_insertf128_ps (t3, u7, 1);
+          __m256 v0 = _mm256_shuffle_ps (t0, t1, 0x44);
+          __m256 v1 = _mm256_shuffle_ps (t0, t1, 0xee);
+          __m256 v2 = _mm256_shuffle_ps (t2, t3, 0x44);
+          __m256 v3 = _mm256_shuffle_ps (t2, t3, 0xee);
+          *(__m256 *)((float *)(&out[4*i+0].data)+outer) = _mm256_shuffle_ps (v0, v2, 0x88);
+          *(__m256 *)((float *)(&out[4*i+1].data)+outer) = _mm256_shuffle_ps (v0, v2, 0xdd);
+          *(__m256 *)((float *)(&out[4*i+2].data)+outer) = _mm256_shuffle_ps (v1, v3, 0x88);
+          *(__m256 *)((float *)(&out[4*i+3].data)+outer) = _mm256_shuffle_ps (v1, v3, 0xdd);
+        }
+      if (remainder > 0 && n_chunks > 0)
+        {
+          // simple re-load all data in the last slot
```

Comment:
The reason why there were two separate code paths for the left-over case depending on whether there were some elements or not was that you re-read some of the elements that were there in the second to last round. In case you did not do an iteration, one would access a negative index.

However, thinking about it I came to the conclusion that it is better to simply use the explicit loop as you suggest. One does save a few operations in the best case (3 more elements to read when there are 4) but loses time in the other cases (1 or 2 elements of 4) because more data gets written, so statistically it is a loss. The good thing is that this saves 150 lines of code :+1: